ldelf/ta_elf.c

```diff
@@ -192,27 +192,37 @@ static void save_hashtab_from_segment(struct ta_elf *elf, unsigned int type,
 	}
 }
 
+static void check_range(struct ta_elf *elf, const char *name, const void *ptr,
+			size_t sz)
+{
+	size_t max_addr = 0;
+
+	if ((vaddr_t)ptr < elf->load_addr)
+		err(TEE_ERROR_GENERIC, "%s %p out of range", name, ptr);
+
+	if (ADD_OVERFLOW((vaddr_t)ptr, sz, &max_addr))
+		err(TEE_ERROR_GENERIC, "%s range overflow", name);
+
+	if (max_addr > elf->max_addr)
+		err(TEE_ERROR_GENERIC,
+		    "%s %p..%#zx out of range", name, ptr, max_addr);
+}
+
 static void check_hashtab(struct ta_elf *elf, void *ptr, size_t num_buckets,
 			  size_t num_chains)
 {
 	size_t num_words = 2;
-	vaddr_t max_addr = 0;
 	size_t sz = 0;
 
-	if ((vaddr_t)ptr < elf->load_addr)
-		err(TEE_ERROR_GENERIC, "Hashtab %p out of range", ptr);
-
 	if (!ALIGNMENT_IS_OK(ptr, uint32_t))
 		err(TEE_ERROR_GENERIC, "Bad alignment of hashtab %p", ptr);
 
 	if (ADD_OVERFLOW(num_words, num_buckets, &num_words) ||
 	    ADD_OVERFLOW(num_words, num_chains, &num_words) ||
-	    MUL_OVERFLOW(num_words, sizeof(uint32_t), &sz) ||
-	    ADD_OVERFLOW((vaddr_t)ptr, sz, &max_addr))
+	    MUL_OVERFLOW(num_words, sizeof(uint32_t), &sz))
 		err(TEE_ERROR_GENERIC, "Hashtab overflow");
 
-	if (max_addr > elf->max_addr)
-		err(TEE_ERROR_GENERIC, "Hashtab %p out of range", ptr);
+	check_range(elf, "Hashtab", ptr, sz);
```

Comment:
very minor: maybe s/Hashtab/hashtab/ here since `dynsymtab` used in lines below.

Reply:
I'll change the dynsymtab strings instead since Hashtab was there first.